File: Simulator/DMSimulator/DMTasks/FirmwareUpdate.c
#include "../Utilities.h"
#include "../SendReport.h"
#include "FirmwareUpdate.h"

// State switch graph: pending -> downloading -> applying -> rebooting -> idle
static DMTaskStep _steps[] =
{
	{ FU_PENDING, 0, FU_DOWNLOADING },
	{ FU_DOWNLOADING, 10, FU_APPLYING },
	{ FU_APPLYING, 10, FU_REBOOTING },
	{ FU_REBOOTING, 10, DM_IDLE },
	{ DM_NULL, 0, DM_NULL }
};

static char* _version = NULL;
static time_t _downloadStartTime = 0;
static time_t _applyStartTime = 0;
static time_t _rebootStartTime = 0;

static BOOL OnEnterState(DMTaskState state, time_t now)
{
	BOOL succeeded = TRUE;

	unsigned char* report = NULL;
	size_t size = 0;

	switch (state)
	{
	case FU_PENDING:
		// No report for entering pending state
		break;

	case FU_DOWNLOADING:
		_downloadStartTime = now;

		if (strcmp(_version, "downloadFail") == 0)
		{
			succeeded = FALSE;

			AllocAndPrintf(
				&report,
				&size,
				"{ 'iothubDM': { 'firmwareUpdate': { 'status': 'download failed', 'log': 'download failed' } } }");
		}
		else
		{
			AllocAndPrintf(
				&report,
				&size,
				"{ 'iothubDM': { 'firmwareUpdate': { 'status': 'downloading', 'log' : 'downloading' } } }");
		}
		break;

	case FU_APPLYING:
		_applyStartTime = now;

		if (strcmp(_version, "applyFail") == 0)
		{
			succeeded = FALSE;

			AllocAndPrintf(
				&report,
				&size,
				"{ 'iothubDM': { 'firmwareUpdate': { 'status': 'apply failed', 'log': 'downloaded(%I64ds) -> apply failed' } } }",
				now - _downloadStartTime);
		}
		else
		{
			AllocAndPrintf(
				&report,
				&size,
				"{ 'iothubDM': { 'firmwareUpdate': { 'status': 'applying', 'log': 'downloaded(%I64ds) -> applying' } } }",
				now - _downloadStartTime);
		}
		break;

	case FU_REBOOTING:
		_rebootStartTime = now;

		if (strcmp(_version, "rebootFail") == 0)
		{
			succeeded = FALSE;

			AllocAndPrintf(
				&report,
				&size,
				"{ 'iothubDM': { 'firmwareUpdate': { 'status': 'reboot failed', 'log': 'downloaded(%I64ds) -> applied(%I64ds) -> reboot failed' } } }",
				_applyStartTime - _downloadStartTime,
				now - _applyStartTime);
		}
		else
		{
			AllocAndPrintf(
				&report,
				&size,
				"{ 'iothubDM': { 'firmwareUpdate': { 'status': 'rebooting', 'log': 'downloaded(%I64ds) -> applied(%I64ds) -> rebooting' } } }",
				_applyStartTime - _downloadStartTime,
				now - _applyStartTime);
		}
		break;

	case DM_IDLE:
		AllocAndPrintf(
			&report,
			&size,
			"{ 'iothubDM': { 'firmwareUpdate': { 'status': 'updated to %s', 'log' : 'downloaded(%I64ds) -> applied(%I64ds) -> rebooted(%I64ds)' } }, 'FirmwareVersion' : '%s' }",
			_version,
			_applyStartTime - _downloadStartTime,
			_rebootStartTime - _applyStartTime,
			now - _rebootStartTime,
			_version);
		break;

	default:
		printf("Unknown FirmwareUpdateState %d", state);
		break;
	}

	if (report != NULL)
	{
		SendReport(report, size);
	}

	return succeeded;
}

static void OnLeaveState(DMTaskState state, time_t now)
{
	unsigned char* report = NULL;
	size_t size = 0;

	switch (state)
	{
	case DM_IDLE:
		// No report for leaving idle state
		break;

	case FU_PENDING:
		// No report for leaving pending state
		break;

	case FU_DOWNLOADING:
		AllocAndPrintf(
			&report,
			&size,
			"{ 'iothubDM': { 'firmwareUpdate': { 'status': 'download completed', 'log': 'downloaded(%I64ds)' } } }",
			now - _downloadStartTime);
		break;

	case FU_APPLYING:
		AllocAndPrintf(
			&report,
			&size,
			"{ 'iothubDM': { 'firmwareUpdate': { 'status': 'apply completed', 'log': 'downloaded(%I64ds) -> applied(%I64ds)' } } }",
			_applyStartTime - _downloadStartTime,
			now - _applyStartTime);
		break;

	case FU_REBOOTING:
		AllocAndPrintf(
			&report,
			&size,
			"{ 'iothubDM': { 'firmwareUpdate': { 'status': 'reboot completed', 'log': 'downloaded(%I64ds) -> applied(%I64ds) -> rebooted(%I64ds)' } } }",
			_applyStartTime - _downloadStartTime,
			_rebootStartTime - _applyStartTime,
			now - _rebootStartTime);
		break;

	default:
		printf("Unknown FirmwareUpdateState %d", state);
		break;
	}

	if (report != NULL)
	{
		SendReport(report, size);
	}
}

BOOL BeginFirmwareUpdate(const char* uri, DMTaskStep** steps, OnEnterStateProc* onEnterStateProc, OnLeaveStateProc* onLeaveStateProc)
{
	// [WORKAROUND] Directly pick the version from the URI
	const char* prefix = "/firmware/";
	char* p = strstr(uri, prefix);
	if (p == NULL)
	{
		return FALSE;
	}
	p += strlen(prefix);

	free(_version);
	_version = _strdup(p);

	*steps = _steps;
	*onEnterStateProc = OnEnterState;
	*onLeaveStateProc = OnLeaveState;
	return TRUE;
}

File: Simulator/Simulator.WebJob/SimulatorCore/Devices/DMTasks/FirmwareUpdate.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Globalization;
using System.Net.Http;
using System.Threading.Tasks;
using Microsoft.Azure.Devices.Applications.RemoteMonitoring.Common.Extensions;
using Microsoft.Azure.Devices.Applications.RemoteMonitoring.Simulator.WebJob.SimulatorCore.Transport;
using Microsoft.Azure.Devices.Client;
using Microsoft.Azure.Devices.Shared;
using Newtonsoft.Json;

namespace Microsoft.Azure.Devices.Applications.RemoteMonitoring.Simulator.WebJob.SimulatorCore.Devices.DMTasks
{
    class FirmwareUpdate : DMTaskBase
    {
        static internal string ReportPrefix = "Method.UpdateFirmware";
        static private string Status = "Status";
        static private string LastUpdate = "LastUpdate";
        static private string Duration = "Duration-s";
        static private string Running = "Running";
        static private string Failed = "Failed";
        static private string Complete = "Complete";

        private Stopwatch _masterWatch;
        private Stopwatch _stepWatch;

        public string Uri { get; private set; }
        public string FirmwareVersion { get; private set; }

        public FirmwareUpdate(MethodRequest request)
        {
            var payload = JsonConvert.DeserializeObject<dynamic>(request.DataAsJson);

            var uri = (string)payload.FwPackageUri;
            if (string.IsNullOrWhiteSpace(uri))
            {
                throw new ArgumentException("Missing FwPackageUri");
            }

            Uri = uri;

            // State switch graph: pending -> downloading -> applying -> rebooting -> idle
            _steps = new List<DMTaskStep>
            {
                new DMTaskStep { CurrentState = DMTaskState.FU_PENDING, ExecuteTime = TimeSpan.Zero, NextState = DMTaskState.FU_DOWNLOADING },
                new DMTaskStep { CurrentState = DMTaskState.FU_DOWNLOADING, ExecuteTime = TimeSpan.FromSeconds(20), NextState = DMTaskState.FU_APPLYING },
                new DMTaskStep { CurrentState = DMTaskState.FU_APPLYING, ExecuteTime = TimeSpan.FromSeconds(20), NextState = DMTaskState.FU_REBOOTING },
                new DMTaskStep { CurrentState = DMTaskState.FU_REBOOTING, ExecuteTime = TimeSpan.FromSeconds(20), NextState = DMTaskState.DM_IDLE }
            };
        }

        protected override async Task<bool> OnEnterStateProc(DMTaskState state, ITransport transport)
        {
            bool succeed = true;
            var report = new TwinCollection();

            switch (state)
            {
                case DMTaskState.FU_PENDING:
                    var clear = new TwinCollection();
                    clear.Set(ReportPrefix, null);
                    await transport.UpdateReportedPropertiesAsync(clear);

                    _masterWatch = Stopwatch.StartNew();
                    _stepWatch = Stopwatch.StartNew();
                    BuildReport(report, Running);
                    break;

                case DMTaskState.FU_DOWNLOADING:
                    _stepWatch = Stopwatch.StartNew();

                    try
                    {
                        using (var client = new HttpClient())
                        {
                            FirmwareVersion = (await client.GetStringAsync(Uri)).Trim();
                        }
                    }
                    catch
                    {
                        succeed = false;
                    }

                    BuildReport(report, "Download", succeed ? Running : Failed);
                    break;

                case DMTaskState.FU_APPLYING:
                    _stepWatch = Stopwatch.StartNew();
                    succeed = FirmwareVersion != "applyFail";
                    BuildReport(report, "Applied", succeed ? Running : Failed);
                    break;

                case DMTaskState.FU_REBOOTING:
                    _stepWatch = Stopwatch.StartNew();
                    succeed = FirmwareVersion != "rebootFail";
                    BuildReport(report, "Reboot", succeed ? Running : Failed);
                    break;

                case DMTaskState.DM_IDLE:
                    BuildReport(report, Complete);

                    report.Set(DeviceBase.StartupTimePropertyName, DateTime.UtcNow.ToString());
                    report.Set(DeviceBase.FirmwareVersionPropertyName, FirmwareVersion);
                    break;

                default:
                    throw new ArgumentOutOfRangeException(nameof(state));
            }

            if (report != null)
            {
                await transport.UpdateReportedPropertiesAsync(report);
            }

            return succeed;
        }

        protected override async Task<bool> OnLeaveStateProc(DMTaskState state, ITransport transport)
        {
            var report = new TwinCollection();

            switch (state)
            {
                case DMTaskState.DM_IDLE:
                    report = null;  // No report for leaving idle state
                    break;

                case DMTaskState.FU_PENDING:
                    report = null;  // No report for leaving pending state
                    break;

                case DMTaskState.FU_DOWNLOADING:
                    BuildReport(report, "Download", Complete);
                    break;

                case DMTaskState.FU_APPLYING:
                    BuildReport(report, "Applied", Complete);
                    break;

                case DMTaskState.FU_REBOOTING:
                    BuildReport(report, "Reboot", Complete);
                    break;

                default:
                    throw new ArgumentOutOfRangeException(nameof(state));
            }

            if (report != null)
            {
                await transport.UpdateReportedPropertiesAsync(report);
            }

            return true;
        }

        private void BuildReport(TwinCollection report, string status)
        {
            report.Set(FormattableString.Invariant($"{ReportPrefix}.{Status}"), status);
            report.Set(FormattableString.Invariant($"{ReportPrefix}.{LastUpdate}"), DateTime.UtcNow.ToString(CultureInfo.CurrentCulture));
            report.Set(FormattableString.Invariant($"{ReportPrefix}.{Duration}"), (int)_masterWatch.Elapsed.TotalSeconds);
        }

        private void BuildReport(TwinCollection report, string stepName, string status)
        {
            report.Set(FormattableString.Invariant($"{ReportPrefix}.{stepName}.{Status}"), status);
            report.Set(FormattableString.Invariant($"{ReportPrefix}.{stepName}.{LastUpdate}"), DateTime.UtcNow.ToString(CultureInfo.CurrentCulture));
            report.Set(FormattableString.Invariant($"{ReportPrefix}.{stepName}.{Duration}"), (int)_stepWatch.Elapsed.TotalSeconds);

            if (status == Failed)
            {
                BuildReport(report, Failed);
            }
            else
            {
                BuildReport(report, Running);
            }
        }
    }
}